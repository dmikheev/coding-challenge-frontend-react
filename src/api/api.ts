interface IApiResponseResult {
  incidents: IIncident[];
  total: number;
}

interface IRawResponse {
  incidents: IIncident[];
}
interface IIncident {
  id: number;
  title: string;
  description: string;
  address: string;
  occurred_at: number;
  updated_at: number;
  source: {
    html_url: string;
  };
  media: {
    image_url_thumb: string;
  };
}

const PROXY_URL = 'https://cors-anywhere.herokuapp.com/';
const ORIGINAL_URL = 'https://bikewise.org:443/api/v2/incidents';

const PER_PAGE = 10;
const INCIDENT_TYPE = 'theft';
const PROXIMITY = 'Berlin, DE';
const PROXIMITY_SQUARE = 50;

function getUrl(page: number): string {
  const params = {
    page,
    per_page: PER_PAGE,
    incident_type: INCIDENT_TYPE,
    proximity: encodeURIComponent(PROXIMITY),
    proximity_square: PROXIMITY_SQUARE,
  };
  const paramsStr = (Object.keys(params) as Array<keyof typeof params>)
    .map((paramName) => `${paramName}=${params[paramName]}`)
    .join('&');

  return `${PROXY_URL}${ORIGINAL_URL}?${paramsStr}`;
}

const api = {
  getIncidents(page = 1): Promise<IApiResponseResult> {
    return fetch(getUrl(page))
      .then((res) => {
        if (!res.ok) {
          throw res;
        }

        return res.json()
          .then((responseJson: IRawResponse) => {
            return {
              incidents: responseJson.incidents,
              total: Number(res.headers.get('total')),
            };
          });
      });
  },
};
export default api;
