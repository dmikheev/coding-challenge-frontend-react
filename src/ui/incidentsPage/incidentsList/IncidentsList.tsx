import { Typography } from '@material-ui/core';
import React from 'react';
import styled from 'styled-components';
import { IGetIncidentsResponseResult } from '../../../api/api';
import IncidentItem from './incidentItem/IncidentItem';

interface IIncidentsListProps {
  className?: string;
  data: IGetIncidentsResponseResult | null;
}
const IncidentsList: React.FC<IIncidentsListProps> = ({ className, data }) => {
  let contentHtml;
  if (data && data.total > 0) {
    const total = (data && data.total) || 0;
    const itemsHtml = data && data.incidents.map((incident) => (
      <StyledIncidentItem key={incident.id} data={incident}/>
    ));

    contentHtml = (
      <React.Fragment>
        <StyledTotalText>Total: {total}</StyledTotalText>
        {itemsHtml}
      </React.Fragment>
    );
  } else if (data) {
    contentHtml = (
      <NoResultsText>
        No results
      </NoResultsText>
    );
  }
  
  return (
    <div className={className}>
      {contentHtml}
    </div>
  );
};
export default IncidentsList;

const StyledTotalText = styled(Typography)`
  margin-bottom: 10px;
  text-align: right;
`;

const StyledIncidentItem = styled(IncidentItem)`
  &:not(:last-child) {
    margin-bottom: 20px;
  }
`;

const NoResultsText = styled(Typography)`
  padding: 10px;
`;
