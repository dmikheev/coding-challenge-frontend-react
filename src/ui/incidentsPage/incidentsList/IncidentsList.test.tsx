import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { IGetIncidentsResponseResult } from '../../../api/api';
import IncidentsList from './IncidentsList';

it('renders correctly', () => {
  const incidentsData: IGetIncidentsResponseResult = {
    incidents: [
      {
        id: 1,
        title: 'Test title 1',
        description: 'Detailed test description 1',
        address: 'Test address 1',
        occurred_at: Date.UTC(2019, 0, 1),
        updated_at: Date.UTC(2019, 1, 1),
        source: {
          html_url: 'https://test.com/bike/1',
        },
        media: {
          image_url_thumb: 'https://test.com/bike/1/photo.jpg',
        },
      },
      {
        id: 2,
        title: 'Test title 2',
        description: 'Detailed test description 2',
        address: 'Test address 2',
        occurred_at: Date.UTC(2019, 0, 2),
        updated_at: Date.UTC(2019, 1, 2),
        source: {
          html_url: 'https://test.com/bike/2',
        },
        media: {
          image_url_thumb: 'https://test.com/bike/2/photo.jpg',
        },
      },
      {
        id: 3,
        title: 'Test title 3',
        description: 'Detailed test description 3',
        address: 'Test address 3',
        occurred_at: Date.UTC(2019, 0, 3),
        updated_at: Date.UTC(2019, 1, 3),
        source: {
          html_url: 'https://test.com/bike/3',
        },
        media: {
          image_url_thumb: 'https://test.com/bike/3/photo.jpg',
        },
      },
    ],
    total: 3,
  };

  const wrapper = shallow(
    <IncidentsList className="test" data={incidentsData}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});
