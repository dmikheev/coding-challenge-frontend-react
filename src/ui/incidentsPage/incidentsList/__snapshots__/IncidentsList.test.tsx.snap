// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="test"
>
  <Styled(WithStyles(ForwardRef(Typography)))>
    Total: 
    3
  </Styled(WithStyles(ForwardRef(Typography)))>
  <Styled(IncidentItem)
    data={
      Object {
        "address": "Test address 1",
        "description": "Detailed test description 1",
        "id": 1,
        "media": Object {
          "image_url_thumb": "https://test.com/bike/1/photo.jpg",
        },
        "occurred_at": 1546300800000,
        "source": Object {
          "html_url": "https://test.com/bike/1",
        },
        "title": "Test title 1",
        "updated_at": 1548979200000,
      }
    }
    key="1"
  />
  <Styled(IncidentItem)
    data={
      Object {
        "address": "Test address 2",
        "description": "Detailed test description 2",
        "id": 2,
        "media": Object {
          "image_url_thumb": "https://test.com/bike/2/photo.jpg",
        },
        "occurred_at": 1546387200000,
        "source": Object {
          "html_url": "https://test.com/bike/2",
        },
        "title": "Test title 2",
        "updated_at": 1549065600000,
      }
    }
    key="2"
  />
  <Styled(IncidentItem)
    data={
      Object {
        "address": "Test address 3",
        "description": "Detailed test description 3",
        "id": 3,
        "media": Object {
          "image_url_thumb": "https://test.com/bike/3/photo.jpg",
        },
        "occurred_at": 1546473600000,
        "source": Object {
          "html_url": "https://test.com/bike/3",
        },
        "title": "Test title 3",
        "updated_at": 1549152000000,
      }
    }
    key="3"
  />
</div>
`;
